File: js/algorithm/dynamic_pro.js
/**
 * 动态规划，代替递归实现的某些场景下的更高效的算法
 */

// 递归实现斐波那契
// 1 1 2 3 5 8 13 ...
function fibonacci(n) {
    if (n == 1 || n == 2) {
        return 1
    }
    return fibonacci(n - 1) + fibonacci(n - 2)
}

// 动态规划实现
function dynamicFibonacci(n) {
    const fibArr = [];
    if (n == 1 || n == 2) {
        return 1
    }
    // 初始化一个长度为n的数组
    for (let i = 0; i < n; i++) {
        fibArr[i] = 0
    }
    fibArr[0] = fibArr[1] = 1
    for (let i = 2; i < n; i++) {
        fibArr[i] = fibArr[i - 1] + fibArr[i - 2]
    }
    return fibArr[n - 1]
}

/**
 * 求解嵌套数组的深度
 * const arr = [1, [2, [3, [4]]], [5, 6], [7, 8], 9]
 * getDepth(arr) // 4
 * 建立状态转移方程
 * 缓存并复用以往的结果
 * 按序从小到大计算
 * 使用递归方法实现
 */
function getDepth(arr) {

    const depArr = []

    for (let i = 0; i < arr.length; i++) {
        depArr[i] = 0
    }

    for (let i = 0; i < arr.length; i++) {
        if (arr[i] instanceof Array) {
            _loopArr(arr[i], i)
        }
    }

    function _loopArr(list, idx) {
        depArr[idx]++
        for (let i = 0; i < list.length; i++) {
            if (list[i] instanceof Array) {
                _loopArr(list[i], idx)
            }
        }
    }

    return Math.max.apply(null, depArr) + 1
}

// 尝试用非递归的方式实现(stack)
function getDepthByStack(arr) {
    const deepRow = [];
    // 初始化存储deepRow的数组, 用于计算每一个元素的数组深度
    arr.forEach(() => {
        deepRow.push(0);
    })

    // 模拟递归用的堆栈
    let loopStack = [];
    for (let i = 0; i < arr.length; i++) {
        loopStack.push(arr[i])
        while (loopStack.length) {
            const curItem = loopStack.pop();
            if (curItem instanceof Array) {
                deepRow[i]++
            }
            for (let j = 0; j < curItem.length; j++) {
                loopStack.push(curItem[j]);
            }
        }
    }

    return Math.max.apply(null, deepRow) + 1
}

module.exports = {
    fibonacci,
    dynamicFibonacci,
    getDepth,
    getDepthByStack
}

File: py/algorithm/dynamic_pro.py
# 备忘录的斐波那契
memo = None


def _fib(n):
    if memo[n] is not None:
        return memo[n]
    if n == 0:
        return 0
    if n == 1:
        return 1
    memo[n] = _fib(n - 1) + _fib(n - 2)
    return memo[n]


def fibonacci(n):
    global memo
    memo = [None] * (n + 1)
    return _fib(n)


# 动态规划实现的斐波那契
def dynamic_fibo(n):
    if n == 1 or n == 2:
        return 1
    fibo_arr = [1, 1] + [None] * (n - 2)
    for item in range(2, n):
        fibo_arr[item] = fibo_arr[item - 1] + fibo_arr[item - 2]
    return fibo_arr[n - 1]


# 充分利用语言特性，使用yield实现
# yield 会把普通函数变成一个generator函数(与js类似, 只是js需要用 * 显式声明 generator)
# 这个函数是需要执行的
def yield_fibo(n):
    # 写法也专业起来了, 开始解构初始化值了，飘了
    a, b = 0, 1
    for item in range(n):
        a, b = b, a + b
        yield a


# 递归扁平化数组
# 如: [1, [2, [3, 4], [5, 6]]] => [1, 2, 3, 4, 5, 6]
def flat_arr(arr):
    flatten = []
    # 判断是否是list
    _flat(arr, lambda item: flatten.append(item))
    return flatten


def _flat(arr, call):
    for item in arr:
        if type(item) == list:
            _flat(item, call)
        else:
            call(item)


# 求数组嵌套的最大深度
# 如[1, [2, [3, [4]], [5, 6], [7, 8, 9]]] => depth: 4
dep_arr = None


def get_depth(arr):
    global dep_arr
    dep_arr = [0] * len(arr)
    for idx, item in enumerate(arr):
        if type(item) == list:
            _add_dep(item, idx)
    return max(dep_arr) + 1


def _add_dep(item, counter_idx):
    global dep_arr
    if type(item) == list:
        dep_arr[counter_idx] += 1
        for sub in item:
            _add_dep(sub, counter_idx)
